src/socket/tcp.rs

```diff
@@ -3062,6 +3082,71 @@ mod test {
         }));
     }
 
+    #[test]
+    fn test_established_retransmit_ack_higher() {
+        // Generate cases where local_seq_no and remote_last_seq wrap when
+        // ACKs are received, both for regular and retransmitted data.
+        for local_seq_start in vec![LOCAL_SEQ, TcpSeqNumber(i32::MAX - 7),
+            TcpSeqNumber(i32::MAX - 6), TcpSeqNumber(i32::MAX - 5), TcpSeqNumber(i32::MAX - 4),
+            TcpSeqNumber(i32::MAX - 3), TcpSeqNumber(i32::MAX - 2)] {
```

Comment:
I don't understand how wrapping is relevant here. TCP sequence numbers are compared modulo 2³². Specifically, `TcpSeqNumber(i32::MAX - 6) - TcpSeqNumber(i32::MAX - 3)` equals `TcpSeqNumber(-6) - TcpSeqNumber(-3)`.

Can you show me a test case that results in an error only when the sequence number wraps? If one exists, it means there is a bug somewhere else apart from the retransmission code.

Reply:
The overflow assertions are fixed in the other commit with an extra test: https://github.com/m-labs/smoltcp/pull/65/commits/3480238359d1d26c7c570af471073049ade002b9 (but are not shown by github anymore since the master merge).
Here the test also includes wrapping and non-wrapping cases to cover the constellations of `local_seq_no`, `old_local_seq_no` and `remote_last_seq`.
When the ACK arrives after the first packet retransmitted for MAX-7 we have: `old_local_seq_no < remote_last_seq < local_seq_no`
For MAX-6 `local_seq_no` already wraps to: `local_seq_no < old_local_seq_no < remote_last_seq`
For MAX-3 `local_seq_no` and `remote_last_seq` now wrapped to: `remote_last_seq < local_seq_no < old_local_seq_no`
In all these cases the value of `remote_last_seq` will be updated.
The last send macro sends an ACK for a normal transmission and the idea was that also the constellations are permuted for this condition where no update should happen. But I just realized that `tx_buffer.len() == 0` is not enough to ensure that this is the case, so I can add some more data to be transfered.

Reply:
Ah, now saw the picked commit, thanks.

Reply:
Expanded the test coverage for normal transmission where remote_last_seq should not be touched.

Reply:
(The example project mentioned in the issue can also be used to reproduce this, but it does not check the contents sent as the test cases do: https://github.com/pothos/taptest)